### PR TITLE
KEYCLOAK-12469 KEYCLOAK-12185 Implement nice design to the screen wit…

### PR DESCRIPTION
**Pull request by mposolda**
…h authentication mechanisms. Add methods to Authentication SPI to retrieve metadata for labels and CSS

### CODE REVIEW

**Comment by mposolda** [comment]
This is how the select-authenticator looks like for the example with having password, OTP and WebAuthn as "1st-factor authenticators" in case that user has at least one credential of each type. 

See details abound this example on slide 5 here - https://docs.google.com/presentation/d/13PpcT26WPTC7v34hS6rj8U63xyZx4lZrPdeHF7qLkdU/edit#slide=id.g70bfb9c9f2_0_0


select-authenticator

**Comment by stianst** [comment]
A bit of nitpicking here:

* The switch user icon is a bit big, not sure it should be blue
* There's too much space at the bottom. Ideal would be to not have the last "----------" and have the same gap from the text to the bottom, as there is on the left side

**Comment by mposolda** [comment]
See the example of new account console UI:
new-account-console
](url)

**Comment by mposolda** [comment]
@stianst For the "nitpicking" issues, I will prefer to create separate JIRA for them and address them later rather than block this PR just because of this. The "username" is already in the Keycloak master and wasn't introduced in this PR. There is design, which was discussed on keycloak-dev and in the keycloak-dev thread, we agreed on this picture https://14121030156225464128.googlegroups.com/attach/56a2122cfc5ce/5.png?part=0.1&view=1&vt=ANaJVrGTi1tppsxSNBijbaq6jobSWNiXxxdkULGmB6cLrOUmrFgvINK7QpxsmsINsJTzKwlyF91xdFaND5jGBwUvo8cpn8u29UFfrWB8jtB6BGT0Dgk7A38 . I know see that "Restart login" icon is different in the design than in current master, however it is still blue.

**Comment by ssilvert** [comment]
I think it would be best to have Password displayed first on the "Signing In" page.  If you make sure that Password is the first category in the list then it will be displayed first.

Also, if I choose to set up a security key, I log in with password and then I am presented with a screen to enter a one-time code.  However, at that point I can not cancel and go back to the account console.  This apparently messed things up on the server.   I manually entered the URL to go back to the account console.  Then when I try to set up Authenticator Application (OTP), I get the flow for security key again.

Also, I tried removing all credentials except password.  Then I started an AIA flow for OTP and canceled it.  At that point, if I try to set up a security key, it immediately sends me back to account console.

So it looks like there are two bugs:
1) We need a cancel button on security key screen.
2) There is a bug in AIA on the server side that causes problems if a setup flow is not completed.

**Comment by mposolda** [comment]
> I think it would be best to have Password displayed first on the "Signing In" page. If you make sure that Password is the first category in the list then it will be displayed first.

Good point, I will try to sort credentials by category (password, then two-factor, then passwordless).

> 
> Also, if I choose to set up a security key, I log in with password and then I am presented with a screen to enter a one-time code. However, at that point I can not cancel and go back to the account console. This apparently messed things up on the server. I manually entered the URL to go back to the account console. Then when I try to set up Authenticator Application (OTP), I get the flow for security key again.
> 
> Also, I tried removing all credentials except password. Then I started an AIA flow for OTP and canceled it. At that point, if I try to set up a security key, it immediately sends me back to account console.
> 
> So it looks like there are two bugs:
> 
>     1. We need a cancel button on security key screen.

Yes, "Cancel" button on the WebAuthn (Security key) setup screen will be good. Created https://issues.redhat.com/browse/KEYCLOAK-12799 . It will need to be done as a follow-up task as there is currently some other WiP on the "WebAuthn setup" screen. Details in https://issues.redhat.com/browse/KEYCLOAK-12178
> 
>     2. There is a bug in AIA on the server side that causes problems if a setup flow is not completed.

I've tried to simulate the scenario you described, but wasn't able to do so. Do you have a chance to create JIRA for that and describe the steps to reproduce? Anyway, I think this will be good as a follow-up task as well

**Comment by ssilvert** [comment]
> I've tried to simulate the scenario you described, but wasn't able to do so. Do you have a chance to create JIRA for that and describe the steps to reproduce? Anyway, I think this will be good as a follow-up task as well

@mposolda Odd thing is, I can't recreate it any more.  However, "Set up Security Key" is not working.  It is sending me "webauthn-register" as the createAction.  Looks like this is not kicking off the AIA so I just get returned to the account console.

**Comment by mposolda** [comment]
> > I've tried to simulate the scenario you described, but wasn't able to do so. Do you have a chance to create JIRA for that and describe the steps to reproduce? Anyway, I think this will be good as a follow-up task as well
> 
> @mposolda Odd thing is, I can't recreate it any more. However, "Set up Security Key" is not working. It is sending me "webauthn-register" as the createAction. Looks like this is not kicking off the AIA so I just get returned to the account console.

I've pushed some updates to the PR. Password is now returned first and two-factor credentials afterwards and passwordless last. More details in the keycloak-dev thread "PR for Account REST Credential API and selecting authentication mechanism"

The fact that you can't setup Security Key may be cause by the fact that you don't have "WebAuthn Register" required action registered in the realm? It is not registered by default and needs to be explicitly registered. See WebAuthn docs for more details.

**Comment by ssilvert** [comment]
> The fact that you can't setup Security Key may be cause by the fact that you don't have "WebAuthn Register" required action registered in the realm? It is not registered by default and needs to be explicitly registered. See WebAuthn docs for more details.

OK.  But in that case, it shouldn't send me anything for WebAuthn.

**Comment by ssilvert** [comment]
> > > I've tried to simulate the scenario you described, but wasn't able to do so. Do you have a chance to create JIRA for that and describe the steps to reproduce? Anyway, I think this will be good as a follow-up task as well
> > 
> > 
> > @mposolda Odd thing is, I can't recreate it any more. However, "Set up Security Key" is not working. It is sending me "webauthn-register" as the createAction. Looks like this is not kicking off the AIA so I just get returned to the account console.
> 
> @ssilvert I've pushed some updates to the PR. Password is now returned first and two-factor credentials afterwards and passwordless last. More details in the keycloak-dev thread "PR for Account REST Credential API and selecting authentication mechanism"
> 
>  See WebAuthn docs for more details.

@mposolda Where do I find the docs?

**Comment by stianst** [comment]
I'm okay with this as is, but there's quite a lot of polish needed.

**Review by ssilvert — DISMISSED** [review]
The part affecting account console looks good, so I'll give my approval.

I'll come back and try it out again if time permits.

**Comment by ssilvert** [comment]
> The part affecting account console looks good, so I'll give my approval.
> 
> I'll come back and try it out again if time permits.

I take that back.  I don't approve yet.  If WebAuthn is not enabled, then the WebAuthn credential type should not be sent to the console.

File: themes/src/main/resources/theme/base/login/select-authenticator.ftl
Comment (vmuzikar) [inline themes/src/main/resources/theme/base/login/select-authenticator.ftl]: If this stays `div`, i.e. JS will be still used, we need to style it to at least look like a link. Currently, mouse cursor doesn't change to "hand" on hover.

**Review by abstractj — COMMENTED** [review]
@mposolda @vmuzikar started a thread about this (sorry about cross posting) and I also added my comments about here: https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/keycloak-dev/u-QzLDxXgqg/u7DlMjZFEQAJ

We may or may not have a bug. Certainly you're the best person to judge if we're testing it correctly.

I will dig more into the code.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
Found a few minor issues.

I must 100% agree with this:
> * The switch user icon is a bit big, not sure it should be blue
> * There's too much space at the bottom. Ideal would be to not have the last "----------" and have the same gap from the text to the bottom, as there is on the left side

Also, found a few (non-related) issues during the review:
https://issues.redhat.com/browse/KEYCLOAK-12817
https://issues.redhat.com/browse/KEYCLOAK-12818

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: This is true and I was also thinking about this. Problem is, that there can be various corner cases. For example:
- Flow with WebAuthn authenticator could be configured as "First broker login" flow
- Flow with WebAuthn authenticator could be configured as "Authetnication flow binding override" for any client (and some of our users has 10K clients in the realm).

 I suppose that when admin configured some flow with WebAuthn authenticator, he actually wants to use WebAuthn somewhere. So I've just used this algorithm as some compromise between accuracy and performance (EG avoid to loop through all the clients and their binding overrides etc). I hope it works as good compromise if we properly document it?

File: themes/src/main/resources/theme/base/login/select-authenticator.ftl
Comment (mposolda) [reply themes/src/main/resources/theme/base/login/select-authenticator.ftl]: @stianst also pointed this. We discuss it will be good to create follow-up JIRA for this to avoid javascript. It will be ideal to do this as a follow-up to not block this PR.

Comment (mposolda) [reply themes/src/main/resources/theme/base/login/select-authenticator.ftl]: Same as previous point. I hope that fine-grained CSS/UI stuff can be addressed as a follow-up

**Comment by mposolda** [comment]
@vmuzikar @ssilvert @abstractj Thanks everyone for the feedback! I need to work a bit more on the PR due the "WebAuthn passwordless" merged in latest master and this PR needs to be adjusted with that and some conflicts will need to be fixed.

Will try to answer some of your concerns/questions.
- WebAuthn docs is at https://www.keycloak.org/docs/latest/server_admin/index.html#_webauthn
- Replied on the issue with WebAuthn userHandle bug on keycloak-dev
- There are lots of minor CSS / UI issues mentioned. We discuss before with @stianst to address those as follow-up. This will allow (among other things). that there will be possibility for more people to work concurrently on issues etc. @vmuzikar Is it fine with you?

- @ssilvert Regarding the concern of WebAuthnRegister required action, there can be still a case when WebAuthnRegister is not registered OR it is disabled, but WebAuthn authenticator is configured for the realm and some users already have WebAuthn security keys registered. I think the correct behaviour in this case will be to still display "Security Key" credential in the account console and all the security keys of the user, but just hide the link "Setup Security Key". We should also make sure that AIA can't be triggered when requiredAction is not registered or is disabled. This is what we discuss yesterday with @vmuzikar and I've tried to mention it also in the thread on keycloak-dev. WDYT? I can update PR in this way. When requiredAction is not registered or it is disabled, I can return null as both "createAction" and "updateAction" and UI will just not display the "Setup Security Key" link. WDYT?

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (vmuzikar) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: @mposolda Thanks for the explanation. That's right and now makes a complete sense.

File: themes/src/main/resources/theme/base/login/select-authenticator.ftl
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/select-authenticator.ftl]: Sounds good. Created https://issues.redhat.com/browse/KEYCLOAK-12824

**Comment by vmuzikar** [comment]
> * There are lots of minor CSS / UI issues mentioned. We discuss before with @stianst to address those as follow-up. This will allow (among other things). that there will be possibility for more people to work concurrently on issues etc. @vmuzikar Is it fine with you?

Sounds good. Created https://issues.redhat.com/browse/KEYCLOAK-12824

File: themes/src/main/resources/theme/base/login/select-authenticator.ftl
Comment (vmuzikar) [reply themes/src/main/resources/theme/base/login/select-authenticator.ftl]: Sounds good. Created https://issues.redhat.com/browse/KEYCLOAK-12823

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (vmuzikar) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: Thinking about this a bit more I believe it might be a problem for OTP. Yes, WebAuthn this approach works but what if admin doesn't want to enable OTP? Users will always see the OTP option in the account console because the OTP authenticator is always present in the built-in browser flow which can't be deleted. Or am I missing something?

**Comment by mposolda** [comment]
@stianst @vmuzikar @ssilvert @abstractj I've rebased the PR with latest master, which means I was able to do add support for "WebAuthn Passwordless" authenticator added to master by the other commit.

I am not sure about the best displayName and helpText for that one. For now, I've used label "Security Key" , which is same label as for "WebAuthn authenticator" . The helpText is bit different - I've used "Use your passwordless security key to log in." . Is it ok or any better ideas?

See the screenshot of "Select authentication" screen and New account console:

passwordless-select-auth

passwordless-account-console

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: Yes. I think in case that OTP authenticator is disabled or it is used in the disabled subflow, it won't be used. That should work fine for built-in browser flow as well, but we need to document it. WDYT?

It will be ideal if there is some way for disable some credential types entirely in some simple way - maybe just switch in admin console or something like that, which will simply allow to enable/disable "otp" or "webauthn" credential etc. But that won't be for 9.0.0...

**Comment by mposolda** [comment]
@stianst @vmuzikar @ssilvert @abstractj Now I see that help text like "Use your passwordless security key to log in." is maybe a bit confusing. As the security key itself is not passwordless - it may require user to provide PIN or anything.

So I am thinking to rather change it to "Use your security key for passwordless log in."

Any better ideas?

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (vmuzikar) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: > Yes. I think in case that OTP authenticator is disabled or it is used in the disabled subflow, it won't be used. That should work fine for built-in browser flow as well, but we need to document it. WDYT? I will need to update PR for this as it currently doesn't take "disabled" into account.

+1
> It will be ideal if there is some way for disable some credential types entirely in some simple way - maybe just switch in admin console or something like that, which will simply allow to enable/disable "otp" or "webauthn" credential etc. But that won't be for 9.0.0...

+1000 :D
However, we might want to wait with that for the new Admin Console where auth flows should be redesigned anyway.

**Comment by ssilvert** [comment]
> @stianst @vmuzikar @ssilvert @abstractj Now I see that help text like "Use your passwordless security key to log in." is maybe a bit confusing. As the security key itself is not passwordless - it may require user to provide PIN or anything.
> 
> So I am thinking to rather change it to "Use your security key for passwordless log in."
> 
> Any better ideas?

I think what you have is clear enough that we can go with it if nobody can think of anything better.

**Comment by abstractj** [comment]
> @stianst @vmuzikar @ssilvert @abstractj Now I see that help text like "Use your passwordless security key to log in." is maybe a bit confusing. As the security key itself is not passwordless - it may require user to provide PIN or anything.
> 
> So I am thinking to rather change it to "Use your security key for passwordless log in."
> 
> Any better ideas?

@mposolda my 2 cents. Let's go with what you suggested, there will always be a room for improvement, and we can do it later. People will certainly provide more feedback about this.

File: services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java
Comment (mposolda) [reply services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java]: I've updated the PR to behave like that and don't consider disabled authenticators. Added test that it is possible to disable OTP and not return it from account REST credential endpoint

**Comment by mposolda** [comment]
@abstractj @vmuzikar @ssilvert I've updated the PR per feedback and discussions. Added separate commit to this PR, so it is easier to re-review. In the end, I am planning to squash this PR to single commit.

Updates are:
- Don't consider disabled authenticators when checking if particular credential type is enabled. For example if OTP Authenticator is disabled in the browser flow (and firstBrokerLogin flow), then OTP credential will be considered disabled as well and not available in the account console. This is per feedback by @vmuzikar .

- Change the helpText of WebAuthn passwordless to "Use your security key for passwordless log in."

- Added some javadoc, especially to account REST endpoints. Hopefully this can be used to "Autogenerate" documentation for REST API

- Don't return "createAction" or "updateAction" from REST endpoint for particular credential type, in case that requiredAction is disabled or not registered. As discussed on the ML. I've updated just REST endpoints for that, but I did not updated Account console UI yet.

It looks UI will need some updates for last point. Now when for example WebAuthn authenticator is enabled, but "WebAuthn Register" required action is disabled, then WebAuthn credentials of the user  are shown in the table, but whole row with displayName, helpText and link for particular credential type is not shown. Looks like just the link should be hidden, but displayName and helpText will be good to show. @ssilvert if I understand correctly what we discuss in the email, you will update UI in some follow-up PR for this?

From my point of view, PR is ready for re-review.

**Comment by ssilvert** [comment]
> * Don't return "createAction" or "updateAction" from REST endpoint for particular credential type, in case that requiredAction is disabled or not registered. As discussed on the ML. I've updated just REST endpoints for that, but I did not updated Account console UI yet.
> 
> It looks UI will need some updates for last point. Now when for example WebAuthn authenticator is enabled, but "WebAuthn Register" required action is disabled, then WebAuthn credentials of the user are shown in the table, but whole row with displayName, helpText and link for particular credential type is not shown. Looks like just the link should be hidden, but displayName and helpText will be good to show. @ssilvert if I understand correctly what we discuss in the email, you will update UI in some follow-up PR for this?
> 

Already working on the UI part.  I can do a PR for it right after this one is merged.

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (ssilvert) [inline server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: Should this be providerID?  For instance, the create action for OTP is "CONFIGURE_TOTP".

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (ssilvert) [inline server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: Same here.  Not really a providerID.  It's the action used to trigger the AIA.  In the case of password, it's "UPDATE_PASSWORD".

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: Yes, "CONFIGURE_TOTP" is the providerID of the required action provider. When requiredAction/AIA is triggered, Keycloak uses the providerID like "CONFIGURE_TOTP" to lookup required action provider - in this case class UpdateTotp. In case of WebAuthn, the name of the action is "webauthn-register" , which is provider id of WebAuthnRegisterFactory.

By default, the alias and providerId of RequiredActionProviderModels attached to the realm is the same, but it may not be necessarily true. Keycloak uses providerId to lookup the requiredAction, not the alias or name of requiredAction.

However if you have any better idea to this javadoc, we can discuss it.

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (mposolda) [reply server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: Yes, the "UPDATE_PASSWORD" is the providerId of requiredAction for update password - it's java class is UpdatePassword. Same case like for "createAction".

However if you have any better idea to this javadoc, we can discuss it.

**Comment by mposolda** [comment]
> > * Don't return "createAction" or "updateAction" from REST endpoint for particular credential type, in case that requiredAction is disabled or not registered. As discussed on the ML. I've updated just REST endpoints for that, but I did not updated Account console UI yet.
> > 
> > It looks UI will need some updates for last point. Now when for example WebAuthn authenticator is enabled, but "WebAuthn Register" required action is disabled, then WebAuthn credentials of the user are shown in the table, but whole row with displayName, helpText and link for particular credential type is not shown. Looks like just the link should be hidden, but displayName and helpText will be good to show. @ssilvert if I understand correctly what we discuss in the email, you will update UI in some follow-up PR for this?
> 
> Already working on the UI part. I can do a PR for it right after this one is merged.

Cool, Thanks!

**Comment by abstractj** [comment]
@mposolda I will do another round of review tomorrow, first thing in the morning. But @ssilvert and @vmuzikar have the final word on this. If they think we are ready to go, let's not hold it.

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (ssilvert) [reply server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: This is fine.  I just didn't remember it being the providerId.  I was thinking the action name was separate.

File: server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java
Comment (ssilvert) [reply server-spi/src/main/java/org/keycloak/credential/CredentialTypeMetadata.java]: NP

**Comment by mposolda** [comment]
> @mposolda I will do another round of review tomorrow, first thing in the morning. But @ssilvert and @vmuzikar have the final word on this. If they think we are ready to go, let's not hold it.

@abstractj @ssilvert @vmuzikar Cool, Thanks! Yes, I hope this is in master by EOD today as there is some work dependent on it (EG documentation, support for dummy credential, which depends on this one and some other PR).

I hope that if there is nothing major found, but just minor things (EG. some HTML/CSS UI stuff or misswording in javadoc), it will be ideal if we can address it as a follow-up. Thanks!

**Review by vmuzikar — APPROVED** [review]
@mposolda Thanks for the update! Found one non-blocking nitpick. Other than that LGTM!

> * Don't consider disabled authenticators when checking if particular credential type is enabled. For example if OTP Authenticator is disabled in the browser flow (and firstBrokerLogin flow), then OTP credential will be considered disabled as well and not available in the account console. This is per feedback by @vmuzikar .

We just must not forget to properly document this as it's not quite intuitive.

Also, found one more non-related issue during the testing: https://issues.redhat.com/browse/KEYCLOAK-12858

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java]: It was reverted already on line 425 in this test.

In general, the AccountRestService test would start to fail if some stuff is not properly reverted :)

Comment (vmuzikar) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java]: @mposolda Ah, right. I missed that. :)

Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/account/AccountRestServiceTest.java]: @vmuzikar Thanks!

> @mposolda Thanks for the update! Found one non-blocking nitpick. Other than that LGTM!
> 
> > * Don't consider disabled authenticators when checking if particular credential type is enabled. For example if OTP Authenticator is disabled in the browser flow (and firstBrokerLogin flow), then OTP credential will be considered disabled as well and not available in the account console. This is per feedback by @vmuzikar .
> 
> We just must not forget to properly document this as it's not quite intuitive.
> 

+1, added comment to https://issues.redhat.com/browse/KEYCLOAK-12753 explicitly about this to not forget when working on docs.

> Also, found one more non-related issue during the testing: https://issues.redhat.com/browse/KEYCLOAK-12858

**Review by abstractj — APPROVED** [review]

**Comment by mposolda** [comment]
@vmuzikar @ssilvert @abstractj Merging, Thanks for the review and feedback. And thanks for all the follow-up bugreports :)

**Comment by mposolda** [comment]
@vmuzikar @ssilvert @abstractj FYI created https://issues.redhat.com/browse/KEYCLOAK-12860 . Hope to have PR for this next week.

**Pull request merged**
